[BUG] `wait_task` raises `TimeoutError` while task is executing

<!--
Thank you for your contribution!

Please review https://github.com/mars-project/mars/blob/master/CONTRIBUTING.rst before opening an issue.
-->

**Describe the bug**
It raised `TimeoutError` on client while the task is running without any errors.

``` Python
mars/services/task/api/web.py:246: in wait_task
    res = await self._request_url(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mars.services.task.api.web.WebTaskAPI object at 0x1752f6070>
method = 'GET'
path = 'http://0.0.0.0:46831/api/session/test_session/task/LC6CDO5j50OiKfJhk5n3Y8au?action=wait&timeout=1'
kwargs = {'request_timeout': 2}, params = {'action': 'wait', 'timeout': '1'}
k = 'timeout', v = '1', url_params = 'action=wait&timeout=1'
path_connector = '?'
request = <tornado.httpclient.HTTPRequest object at 0x1752a1550>

    async def _request_url(self, method, path, **kwargs):
        self._running_loop = asyncio.get_running_loop()
    
        if "data" in kwargs:
            kwargs["body"] = kwargs.pop("data")
    
        if "params" in kwargs:
            params = kwargs.pop("params")
            for k, v in params.items():
                if isinstance(v, (list, tuple, set)):
                    params[k] = ",".join(str(i) for i in v)
            url_params = urllib.parse.urlencode(params)
            path_connector = "?" if "?" not in path else "&"
            path += path_connector + url_params
    
        try:
            request = HTTPRequest(path, method=method, **kwargs)
            if self.request_rewriter:
                request = self.request_rewriter(request)
            res = await self._client.fetch(request, raise_error=False)
        except HTTPTimeoutError as ex:
>           raise TimeoutError(str(ex)) from None
E           TimeoutError: Timeout during request

mars/services/web/core.py:239: TimeoutError
```

--- a/mars/services/task/api/web.py
+++ b/mars/services/task/api/web.py
@@ -14,7 +14,11 @@
 
 import base64
 import json
+from datetime import timedelta
 from typing import Callable, List, Optional, Union
+
+from tornado import gen
+from tornado.util import TimeoutError
 
 from ....core import TileableGraph, Tileable
 from ....utils import serialize_serializable, deserialize_serializable
@@ -152,8 +156,17 @@ class TaskWebAPIHandler(MarsServiceWebAPIHandler):
         timeout = self.get_argument("timeout", None) or None
         timeout = float(timeout) if timeout is not None else None
         oscar_api = await self._get_oscar_task_api(session_id)
-        res = await oscar_api.wait_task(task_id, timeout)
-        self.write(json.dumps(_json_serial_task_result(res)))
+        if timeout:
+            try:
+                res = await gen.with_timeout(
+                    timedelta(seconds=timeout), oscar_api.wait_task(task_id, timeout)
+                )
+                self.write(json.dumps(_json_serial_task_result(res)))
+            except TimeoutError:
+                self.write(json.dumps({}))
+        else:
+            res = await oscar_api.wait_task(task_id, timeout)
+            self.write(json.dumps(_json_serial_task_result(res)))
 
     @web_api("(?P<task_id>[^/]+)", method="delete")
     async def cancel_task(self, session_id: str, task_id: str):
@@ -237,7 +250,7 @@ class WebTaskAPI(AbstractTaskAPI, MarsWebAPIClientMixin):
     async def wait_task(self, task_id: str, timeout: float = None):
         path = f"{self._address}/api/session/{self._session_id}/task/{task_id}"
         # client timeout should be longer than server timeout.
-        server_timeout = "" if timeout is None else str(max(timeout / 2.0, timeout - 1))
+        server_timeout = "" if timeout is None else str(int(timeout / 2.0))
         params = {"action": "wait", "timeout": server_timeout}
         res = await self._request_url(
             "GET", path, params=params, request_timeout=timeout or 0

--- a/mars/services/task/api/web.py
+++ b/mars/services/task/api/web.py
@@ -18,7 +18,7 @@ from datetime import timedelta
 from typing import Callable, List, Optional, Union
 
 from tornado import gen
-from tornado.util import TimeoutError
+from tornado.util import TimeoutError as TornadoTimeout
 
 from ....core import TileableGraph, Tileable
 from ....utils import serialize_serializable, deserialize_serializable
@@ -162,7 +162,7 @@ class TaskWebAPIHandler(MarsServiceWebAPIHandler):
                     timedelta(seconds=timeout), oscar_api.wait_task(task_id, timeout)
                 )
                 self.write(json.dumps(_json_serial_task_result(res)))
-            except TimeoutError:
+            except TornadoTimeout:
                 self.write(json.dumps({}))
         else:
             res = await oscar_api.wait_task(task_id, timeout)

--- a/mars/services/task/api/web.py
+++ b/mars/services/task/api/web.py
@@ -12,13 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import asyncio
 import base64
 import json
-from datetime import timedelta
 from typing import Callable, List, Optional, Union
-
-from tornado import gen
-from tornado.util import TimeoutError as TornadoTimeout
 
 from ....core import TileableGraph, Tileable
 from ....utils import serialize_serializable, deserialize_serializable
@@ -158,11 +155,12 @@ class TaskWebAPIHandler(MarsServiceWebAPIHandler):
         oscar_api = await self._get_oscar_task_api(session_id)
         if timeout:
             try:
-                res = await gen.with_timeout(
-                    timedelta(seconds=timeout), oscar_api.wait_task(task_id, timeout)
+                res = await asyncio.wait_for(
+                    asyncio.shield(oscar_api.wait_task(task_id, timeout)),
+                    timeout=timeout,
                 )
                 self.write(json.dumps(_json_serial_task_result(res)))
-            except TornadoTimeout:
+            except asyncio.TimeoutError:
                 self.write(json.dumps({}))
         else:
             res = await oscar_api.wait_task(task_id, timeout)
@@ -250,7 +248,7 @@ class WebTaskAPI(AbstractTaskAPI, MarsWebAPIClientMixin):
     async def wait_task(self, task_id: str, timeout: float = None):
         path = f"{self._address}/api/session/{self._session_id}/task/{task_id}"
         # client timeout should be longer than server timeout.
-        server_timeout = "" if timeout is None else str(int(timeout / 2.0))
+        server_timeout = "" if timeout is None else str(max(timeout / 2.0, timeout - 1))
         params = {"action": "wait", "timeout": server_timeout}
         res = await self._request_url(
             "GET", path, params=params, request_timeout=timeout or 0

--- a/mars/services/task/supervisor/manager.py
+++ b/mars/services/task/supervisor/manager.py
@@ -34,7 +34,7 @@ from ..config import task_options
 from ..core import Task, new_task_id, TaskStatus
 from ..errors import TaskNotExist
 from .preprocessor import TaskPreprocessor
-from .processor import TaskProcessorActor
+from .processor import TaskProcessorActor, TaskProcessor
 
 logger = logging.getLogger(__name__)
 
@@ -43,15 +43,18 @@ class TaskConfigurationActor(mo.Actor):
     def __init__(
         self,
         task_conf: Dict[str, Any],
+        task_processor_cls: Type[TaskProcessor] = None,
         task_preprocessor_cls: Type[TaskPreprocessor] = None,
     ):
         for name, value in task_conf.items():
             setattr(task_options, name, value)
+        self._task_processor_cls = task_processor_cls
         self._task_preprocessor_cls = task_preprocessor_cls
 
     def get_config(self):
         return {
             "task_options": task_options,
+            "task_processor_cls": self._task_processor_cls,
             "task_preprocessor_cls": self._task_preprocessor_cls,
         }
 
@@ -77,6 +80,7 @@ class TaskManagerActor(mo.Actor):
         self._session_id = session_id
 
         self._config = None
+        self._task_processor_cls = None
         self._task_preprocessor_cls = None
         self._last_idle_time = None
 
@@ -104,9 +108,10 @@ class TaskManagerActor(mo.Actor):
             TaskConfigurationActor.default_uid(), address=self.address
         )
         task_conf = await configuration_ref.get_config()
-        self._config, self._task_preprocessor_cls = (
+        self._config, self._task_preprocessor_cls, self._task_processor_cls = (
             task_conf["task_options"],
             task_conf["task_preprocessor_cls"],
+            task_conf["task_processor_cls"],
         )
         self._task_preprocessor_cls = self._get_task_preprocessor_cls()
 
@@ -159,6 +164,7 @@ class TaskManagerActor(mo.Actor):
                 self._session_id,
                 parent_task_id,
                 task_name=task_name,
+                task_processor_cls=self._task_processor_cls,
                 address=self.address,
                 uid=uid,
             )

--- a/mars/services/task/supervisor/processor.py
+++ b/mars/services/task/supervisor/processor.py
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 import asyncio
+import importlib
 import itertools
 import logging
 import operator
@@ -525,11 +526,18 @@ class TaskProcessorActor(mo.Actor):
     _processed_task_ids: Set[str]
     _cur_processor: Optional[TaskProcessor]
 
-    def __init__(self, session_id: str, task_id: str, task_name: str = None):
+    def __init__(
+        self,
+        session_id: str,
+        task_id: str,
+        task_name: str = None,
+        task_processor_cls: Type[TaskPreprocessor] = None,
+    ):
         self.session_id = session_id
         self.task_id = task_id
         self.task_name = task_name
 
+        self._task_processor_cls = self._get_task_processor_cls(task_processor_cls)
         self._task_id_to_processor = dict()
         self._cur_processor = None
         self._subtask_decref_events = dict()
@@ -559,7 +567,7 @@ class TaskProcessorActor(mo.Actor):
         task_preprocessor = task_preprocessor_cls(
             task, tiled_context=tiled_context, config=config
         )
-        processor = TaskProcessor(
+        processor = self._task_processor_cls(
             task,
             task_preprocessor,
             self._cluster_api,
@@ -571,6 +579,15 @@ class TaskProcessorActor(mo.Actor):
 
         # tell self to start running
         await self.ref().start.tell()
+
+    @classmethod
+    def _get_task_processor_cls(cls, task_processor_cls):
+        if task_processor_cls is not None:
+            assert isinstance(task_processor_cls, str)
+            module, name = task_processor_cls.rsplit(".", 1)
+            return getattr(importlib.import_module(module), name)
+        else:
+            return TaskProcessor
 
     def _get_unprocessed_task_processor(self):
         for processor in self._task_id_to_processor.values():

--- a/mars/services/task/supervisor/service.py
+++ b/mars/services/task/supervisor/service.py
@@ -37,10 +37,12 @@ class TaskSupervisorService(AbstractService):
     async def start(self):
         task_config = self._config.get("task", dict())
         options = task_config.get("default_config", dict())
+        task_processor_cls = task_config.get("task_processor_cls")
         task_preprocessor_cls = task_config.get("task_preprocessor_cls")
         await mo.create_actor(
             TaskConfigurationActor,
             options,
+            task_processor_cls=task_processor_cls,
             task_preprocessor_cls=task_preprocessor_cls,
             address=self._address,
             uid=TaskConfigurationActor.default_uid(),

--- a/mars/services/task/tests/test_service.py
+++ b/mars/services/task/tests/test_service.py
@@ -32,6 +32,7 @@ from ...subtask import SubtaskStatus
 from ...web import WebActor
 from ...meta import MetaAPI
 from .. import TaskAPI, TaskStatus, WebTaskAPI
+from ..supervisor.processor import TaskProcessor
 from ..errors import TaskNotExist
 
 
@@ -45,7 +46,7 @@ async def actor_pools():
                 subprocess_start_method="spawn",
             )
         else:
-            kw = dict(n_process=0, subprocess_start_method="spawn")
+            kw = dict(n_process=1, subprocess_start_method="spawn")
         pool = await mo.create_actor_pool("127.0.0.1", **kw)
         await pool.start()
         return pool
@@ -57,11 +58,9 @@ async def actor_pools():
         await asyncio.gather(sv_pool.stop(), worker_pool.stop())
 
 
-@pytest.mark.parametrize(indirect=True)
-@pytest.fixture(params=[False, True])
-async def start_test_service(actor_pools, request):
-    sv_pool, worker_pool = actor_pools
-
+async def _start_services(
+    supervisor_pool, worker_pool, request, task_processor_cls=None
+):
     config = {
         "services": [
             "cluster",
@@ -75,28 +74,31 @@ async def start_test_service(actor_pools, request):
         ],
         "cluster": {
             "backend": "fixed",
-            "lookup_address": sv_pool.external_address,
+            "lookup_address": supervisor_pool.external_address,
             "resource": {"numa-0": Resource(num_cpus=2)},
         },
         "meta": {"store": "dict"},
         "scheduling": {},
         "task": {},
     }
+    if task_processor_cls:
+        config["task"]["task_processor_cls"] = task_processor_cls
     if request:
         config["services"].append("web")
-
-    await start_services(NodeRole.SUPERVISOR, config, address=sv_pool.external_address)
+    await start_services(
+        NodeRole.SUPERVISOR, config, address=supervisor_pool.external_address
+    )
     await start_services(NodeRole.WORKER, config, address=worker_pool.external_address)
 
     session_id = "test_session"
-    session_api = await SessionAPI.create(sv_pool.external_address)
+    session_api = await SessionAPI.create(supervisor_pool.external_address)
     await session_api.create_session(session_id)
 
     if not request.param:
-        task_api = await TaskAPI.create(session_id, sv_pool.external_address)
+        task_api = await TaskAPI.create(session_id, supervisor_pool.external_address)
     else:
         web_actor = await mo.actor_ref(
-            WebActor.default_uid(), address=sv_pool.external_address
+            WebActor.default_uid(), address=supervisor_pool.external_address
         )
         web_address = await web_actor.get_web_address()
         task_api = WebTaskAPI(session_id, web_address)
@@ -104,8 +106,17 @@ async def start_test_service(actor_pools, request):
     assert await task_api.get_task_results() == []
 
     # create mock meta and storage APIs
-    _ = await MetaAPI.create(session_id, sv_pool.external_address)
+    _ = await MetaAPI.create(session_id, supervisor_pool.external_address)
     storage_api = await MockStorageAPI.create(session_id, worker_pool.external_address)
+    return task_api, storage_api, config
+
+
+@pytest.mark.parametrize(indirect=True)
+@pytest.fixture(params=[False, True])
+async def start_test_service(actor_pools, request):
+    sv_pool, worker_pool = actor_pools
+
+    task_api, storage_api, config = await _start_services(sv_pool, worker_pool, request)
 
     try:
         yield sv_pool.external_address, task_api, storage_api
@@ -113,6 +124,66 @@ async def start_test_service(actor_pools, request):
         await MockStorageAPI.cleanup(worker_pool.external_address)
         await stop_services(NodeRole.WORKER, config, worker_pool.external_address)
         await stop_services(NodeRole.SUPERVISOR, config, sv_pool.external_address)
+
+
+class MockTaskProcessor(TaskProcessor):
+    @classmethod
+    def _get_decref_stage_chunk_keys(cls, stage_processor):
+        import time
+
+        # time.sleep to block async thread
+        time.sleep(5)
+        return super()._get_decref_stage_chunk_keys(stage_processor)
+
+
+@pytest.mark.parametrize(indirect=True)
+@pytest.fixture(params=[True])
+async def start_test_service_with_mock(actor_pools, request):
+    sv_pool, worker_pool = actor_pools
+
+    task_api, storage_api, config = await _start_services(
+        sv_pool,
+        worker_pool,
+        request,
+        task_processor_cls="mars.services.task.tests.test_service.MockTaskProcessor",
+    )
+
+    try:
+        yield sv_pool.external_address, task_api, storage_api
+    finally:
+        await MockStorageAPI.cleanup(worker_pool.external_address)
+        await stop_services(NodeRole.WORKER, config, worker_pool.external_address)
+        await stop_services(NodeRole.SUPERVISOR, config, sv_pool.external_address)
+
+
+@pytest.mark.asyncio
+async def test_task_timeout_execution(start_test_service_with_mock):
+    _sv_pool_address, task_api, storage_api = start_test_service_with_mock
+
+    def f1():
+        return np.arange(5)
+
+    def f2():
+        return np.arange(5, 10)
+
+    def f3(f1r, f2r):
+        return np.concatenate([f1r, f2r]).sum()
+
+    r1 = mr.spawn(f1)
+    r2 = mr.spawn(f2)
+    r3 = mr.spawn(f3, args=(r1, r2))
+
+    graph = TileableGraph([r3.data])
+    next(TileableGraphBuilder(graph).build())
+
+    task_id = await task_api.submit_tileable_graph(graph, fuse_enabled=False)
+    assert await task_api.get_last_idle_time() is None
+    assert isinstance(task_id, str)
+
+    await task_api.wait_task(task_id, timeout=2)
+    task_result = await task_api.get_task_result(task_id)
+
+    assert task_result.status == TaskStatus.terminated
 
 
 @pytest.mark.asyncio

